File: cpp/src/layout.cpp
#include <iostream>
#include <layout.h>

namespace rbf
{

    Layout::Layout(string xml_file, size_t initial_size)
    {
        // save file name for future use
        _xml_file = xml_file;

        // open document
        xml_document doc;
        auto result = doc.load_file(_xml_file.data());

        // test if loading is successful
        if (!result)
        {
            throw runtime_error("unable to open file!");
        }

        // root node is rbfile
        auto root = doc.child("rbfile");

        // lookup for field types
        unordered_map<string, FieldType> ftype_map;

        for (auto node: root.children("fieldtype"))
        {
            string data_representation(node.attribute("name").value());
            string data_description(node.attribute("type").value());

            // add field type in our map to later refer to them
            ftype_map.insert(pair<string, FieldType>(data_representation, FieldType(data_representation, data_description)));
        }

        // now look up records
        for (auto node: root.children("record"))
        {
            string rec_name(node.attribute("name").value());
            string rec_desc(node.attribute("description").value());

            // create record and add it to our map
            auto p_rec = make_unique<Record>(rec_name, rec_desc);
            _record_map.insert(pair<string, RecordPtr>(rec_name, move(p_rec)));

            // pre-allocate record size
            _record_map[rec_name]->reserve(initial_size);

            // look up fields and add them to record
            for (auto field: node.children("field"))
            {
                string field_name(field.attribute("name").value());
                string field_desc(field.attribute("description").value());
                string field_type(field.attribute("type").value());
                string field_length(field.attribute("length").value());

                // get saved FieldType object
                auto ft = ftype_map[field_type];

                // add Field to last record created
                auto f = Field(field_name, field_desc, ftype_map[field_type], stoul(field_length));
                _record_map[rec_name]->push_back(f);
            }
        }

    }
}

File: ruby/lib/layout.rb
require 'rexml/document'
include REXML

# This class is a representation of all records and classes composing a record-based file.
# As a convention, the layout is defined is an XML file, with the following convention:
#
# * root tag is <rbfile>
# * field type tag (<fieldtype> XML tag)
# * record tag with the following attributes: name & description (<record> XML tag)
# * field tag inside a record tag (<field> XML tag)
#
# Suppose you get the following layout file (world_data.xml) describing a record-based file
# listing continent data:
#
#   <?xml version="1.0" encoding="UTF-8"?>
#   <!-- inspired from https://en.wikipedia.org/wiki/List_of_continents_by_GDP_%28nominal%29 -->
#   <!-- and http://www.nationsonline.org/oneworld/asia.htm -->
#   <rbfile
#       xmlns="http://www.w3schools.com"
#       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
#       xsi:schemaLocation="http://www.w3schools.com rbf.xsd"
#   >
#
#       <meta version="1.0" description="Continents, countries, cities" ignoreLine="^#" skipField="ID" mapper="type:1 map:0..4"/>
#
#       <fieldtype name="CHAR" type="string" pattern="\w+" format=""/>
#       <fieldtype name="NUM" type="decimal"/>
#       <fieldtype name="INT" type="integer"/>
#
#       <record name="CONT" description="Continent data">
#           <field name="ID" description="Record ID" length="4" type="CHAR"/>
#           <field name="NAME" description="Name of the continent" length="15" type="CHAR"/>
#           <field name="AREA" description="Area of the continent" length="20" type="NUM"/>
#           <field name="POPULATION" description="Population of the continent" length="20" type="NUM"/>
#           <field name="DENSITY" description="Density per km2" length="9" type="NUM"/>
#           <field name="CITY" description="Most populus city" length="20" type="CHAR"/>
#       </record>
#
#       <record name="COUN" description="Country data">
#           <field name="ID" description="Record ID" length="4" type="CHAR"/>
#           <field name="NAME" description="Name of the country" length="30" type="CHAR"/>
#           <field name="POPULATION" description="Number of inhabitants" length="20" type="INT"/>
#           <field name="CAPITAL" description="Capital of the country" length="20" type="CHAR"/>
#       </record>
#
#   </rbfile>
#
# ==== Example
#
#      # create a layout object from XML definition file layout.xml
#      layout = Layout.new("world_data.xml")
#
#      # get continent data record
#      cont = layout["CONT"]
#      cont.size                #=> 6
#      cont[0].type.type        #=> :string
#
#  
class Layout < Element
    include Enumerable

    # hash of records, key is record name, value is Record object
    attr_reader :rmap

    # build fields & records from XML layout file
    # :call-seq:
    # new(xml_file) -> new_Layout
    def initialize(xml_file)
        # check file existence
        raise ArgumentError, "xml layout file #{xml_file} is not found!" unless File.exist?(xml_file)

        # initialize hash of FieldType objects
        ft_map = Hash.new

        # open document
        doc = Document.new(File.new(xml_file))
        root = doc.root

        # call parent ctor
        super(xml_file, root.attributes["description"], root.attributes["reclength"])

        # initialize record list
        @rmap = Hash.new

        # create all field types
        doc.elements.each("rbfile/fieldtype") {|ft|
            ft_map[ft.attributes["name"]] = FieldType.new(ft.attributes["name"], ft.attributes["type"])
        }

        # read record tags
        doc.elements.each("rbfile/record") {|e| 
            # create record
            recname = e.attributes["name"]
            @rmap[recname] = Record.new(recname, e.attributes["description"])

            # now add fields
            e.elements.each {|f| 
                field_type = ft_map[f.attributes["type"]]
                @rmap[recname] << 
                    Field.new(f.attributes["name"], f.attributes["description"], field_type, 
                              f.attributes["length"].to_i)
            }          
        }
    end

    # iterator on records
    def each(&block)
        @rmap.each_value(&block)
    end

    # access to the Record object matching record name
    # :call-seq:
    # [](recname) -> Record object
    def [](recname)
        @rmap[recname]
    end

end



